### ISSUE TITLE
BUG: SciPy 1.13.0 not buildable on old macOS due to pocketfft code

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#20300]
This occurred while [building](https://github.com/conda-forge/scipy-feedstock/pull/270) rc1 in conda-forge and is exactly the same issue as https://github.com/numpy/numpy/issues/25940, but I'm raising it for release-tracking purposes, as - aside from fixing it in some way - we also need to update the submodule. Failure looks like:

```
[881/1477] Compiling C++ object scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o
FAILED: scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o 
x86_64-apple-darwin13.4.0-clang++ -Iscipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p -Iscipy/fft/_pocketfft -I../scipy/fft/_pocketfft -I../scipy/_lib/pocketfft -I$PREFIX/include -I$PREFIX/include/python3.10 -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c++17 -O3 -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/scipy-split-1.13.0rc1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -mmacosx-version-min=10.9 -DPOCKETFFT_PTHREADS -DPOCKETFFT_CACHE_SIZE=16 -MD -MQ scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o -MF scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o.d -o scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o -c ../scipy/fft/_pocketfft/pypocketfft.cxx
In file included from ../scipy/fft/_pocketfft/pypocketfft.cxx:19:
../scipy/_lib/pocketfft/pocketfft_hdronly.h:160:15: error: no member named 'aligned_alloc' in the global namespace; did you mean simply 'aligned_alloc'?
  void *ptr = ::aligned_alloc(align,(size+align-1)&(~(align-1)));
              ^~~~~~~~~~~~~~~
              aligned_alloc
../scipy/_lib/pocketfft/pocketfft_hdronly.h:157:14: note: 'aligned_alloc' declared here
inline void *aligned_alloc(size_t align, size_t size)
             ^
1 error generated.
```

Quoting myself from the numpy issue:
> Looking at the [code](https://github.com/mreineck/pocketfft/blob/0f7aa1225b065938fc263b7914df16b8c1cbc9d7/pocketfft_hdronly.h#L155-L160) in question
> 
> ```c
> // the __MINGW32__ part in the conditional below works around the problem that
> // the standard C++ library on Windows does not provide aligned_alloc() even
> // though the MinGW compiler and MSVC may advertise C++17 compliance.
> // aligned_alloc is only supported from MacOS 10.15.
> #if (__cplusplus >= 201703L) && (!defined(__MINGW32__)) && (!defined(_MSC_VER)) && (MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_15)
> inline void *aligned_alloc(size_t align, size_t size)
> ```
> 
> there's already a guard to only use this on macOS >=10.15, but it looks like `MAC_OS_X_VERSION_MIN_REQUIRED` might not be defined (correctly).

**Comment by lucascolley** [comment]
which version of OSX does it fail on?

**Comment by h-vetinari** [comment]
10.9. we could bump to 10.13, but in either case the detection and preprocessor-guard is not working

**Comment by tylerjereddy** [comment]
@h-vetinari tried to sync in the small preprocessor directive patch to `conda-forge` without success, so this issue should remain open and we should probably just disable the alloc thing completely then, as upstream did and NumPy likely will.

**Comment by h-vetinari** [comment]
This still occurs with SciPy 1.13.0, with https://github.com/scipy/pocketfft/commit/0bf2b51224e7c4e973724bac67d2577c7831e27c, because the submodule in scipy hasn't been updated to incorporate https://github.com/scipy/pocketfft/pull/3 yet.
```
[880/1475] Compiling C++ object scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o
FAILED: scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o 
x86_64-apple-darwin13.4.0-clang++ -Iscipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p -Iscipy/fft/_pocketfft -I../scipy/fft/_pocketfft -I../scipy/_lib/pocketfft -I$PREFIX/include -I$PREFIX/include/python3.10 -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c++17 -O3 -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/scipy-split-1.13.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -mmacosx-version-min=10.9 -DPOCKETFFT_PTHREADS -DPOCKETFFT_CACHE_SIZE=16 -MD -MQ scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o -MF scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o.d -o scipy/fft/_pocketfft/pypocketfft.cpython-310-darwin.so.p/pypocketfft.cxx.o -c ../scipy/fft/_pocketfft/pypocketfft.cxx
In file included from ../scipy/fft/_pocketfft/pypocketfft.cxx:19:
../scipy/_lib/pocketfft/pocketfft_hdronly.h:163:15: error: no member named 'aligned_alloc' in the global namespace; did you mean simply 'aligned_alloc'?
  void *ptr = ::aligned_alloc(align,(size+align-1)&(~(align-1)));
              ^~~~~~~~~~~~~~~
              aligned_alloc
../scipy/_lib/pocketfft/pocketfft_hdronly.h:160:14: note: 'aligned_alloc' declared here
inline void *aligned_alloc(size_t align, size_t size)
             ^
1 error generated.
```

**Comment by lucascolley** [comment]
looks like gh-20333 missed a backport @tylerjereddy ?

**Comment by h-vetinari** [comment]
The fix in [conda-forge](https://github.com/conda-forge/scipy-feedstock/pull/271) is to simply pull in https://github.com/scipy/pocketfft/commit/9367142748fcc9696a1c9e5a99b76ed9897c9daa, but this should still be fixed on the 1.13 branch

**Comment by tylerjereddy** [comment]
It didn't have a backport label yeah, I've added it. This will have to wait for `1.13.1` though, and I'm not sure when that will be since the bandwidth requirements to get `1.13.0` out quickly were fairly high. We'll see how many complaints we get.

**Comment by doronbehar** [comment]
We also experience this on NixOS. It was a bit hard to apply https://github.com/scipy/pocketfft/commit/9367142748fcc9696a1c9e5a99b76ed9897c9daa on our git checkout because it is a submodule, but still I think we are OK now.

**Comment by h-vetinari** [comment]
This was fixed and backported to 1.13.1, so closing.
